module/state_synchronization/execution_data_service.go

```diff
@@ -0,0 +1,280 @@
+package state_synchronization
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ipfs/go-cid"
+
+	"github.com/onflow/flow-go/model/encoding"
+	"github.com/onflow/flow-go/network"
+)
+
+const (
+	defaultMaxBlobSize      = 1 << 20 // 1MiB
+	defaultMaxBlobTreeDepth = 2       // prevents malicious CID from causing download of unbounded amounts of data
+	defaultBlobBatchSize    = 16
+)
+
+var ErrBlobTreeDepthExceeded = errors.New("blob tree depth exceeded")
+
+// ExecutionDataService handles adding and getting execution data from a blobservice
+type ExecutionDataService struct {
+	serializer  *serializer
+	blobService network.BlobService
+	maxBlobSize int
+}
+
+func NewExecutionDataService(
+	codec encoding.Codec,
+	compressor network.Compressor,
+	blobService network.BlobService,
+) *ExecutionDataService {
+	return &ExecutionDataService{&serializer{codec, compressor}, blobService, defaultMaxBlobSize}
+}
+
+func (s *ExecutionDataService) storeBatch(ctx context.Context, br *BlobReceiver) ([]network.Blob, error) {
+	var blobs []network.Blob
+	var err error
+
+	for i := 0; i < defaultBlobBatchSize; i++ {
+		var blob network.Blob
+		blob, err = br.Receive()
+
+		if err != nil {
+			break
+		}
+
+		blobs = append(blobs, blob)
+	}
+
+	return blobs, err
+}
+
+func (s *ExecutionDataService) storeBlobs(parent context.Context, br *BlobReceiver) ([]cid.Cid, error) {
+	defer br.Close()
+
+	ctx, cancel := context.WithCancel(parent)
```

Comment:
So that when we return from this function, we cancel any ongoing `AddBlobs` requests that returned with an error.

It's possible that it isn't actually necessary, because `AddBlobs` may properly handle canceling all goroutines before returning an error, but just wanted to be safe.

Reply:
ok, can you describe that in a comment here then